source/robot_model/src/Model.cpp

```diff
@@ -29,17 +49,145 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+                                std::istringstream(urdf),
+                                pinocchio::COLLISION, 
+                                this->geom_model_, 
+                                package_paths);
+      this->geom_model_.addAllCollisionPairs();
+    
+      std::vector<pinocchio::CollisionPair> excluded_pairs = this->generate_joint_exclusion_list();
+
+      // remove collision pairs for linked joints (i.e. parent-child joints)
+      for (const auto& pair : excluded_pairs) {
+          this->geom_model_.removeCollisionPair(pair);
+      }
+
+      this->geom_data_ = pinocchio::GeometryData(this->geom_model_);
+    } catch (const std::exception& ex) {
+        throw robot_model::exceptions::CollisionGeometryException("Failed to initialize geometry model for " + this->get_robot_name() + ": " + ex.what());
+    }
+}
+
+std::vector<pinocchio::CollisionPair> Model::generate_joint_exclusion_list() {
+    std::vector<pinocchio::CollisionPair> excluded_pairs;
+    // Iterate through all joints, except the universe joint (0), which has no parent
+    for (pinocchio::JointIndex joint_id = 1u; joint_id < static_cast<pinocchio::JointIndex>(this->robot_model_.njoints); ++joint_id){
+        // Find the parent joint of the current joint
+        pinocchio::JointIndex parent_id = this->robot_model_.parents[joint_id];
+        
+        // TODO: Replace this logic with actual geometry index lookup
+        auto getGeometryIndexForJoint = [](pinocchio::JointIndex joint_id) -> int {
+            return static_cast<int>(joint_id);
+        };
+
+        int geometryIndex1 = getGeometryIndexForJoint(joint_id);
+        int geometryIndex2 = getGeometryIndexForJoint(parent_id);
+
+        // Check if the geometry indices are not equal
+        if (geometryIndex1 != geometryIndex2) {
+            excluded_pairs.push_back(pinocchio::CollisionPair(geometryIndex2, geometryIndex1));
+        }
+    }
+    return excluded_pairs;
+}
+
+unsigned int Model::get_number_of_collision_pairs() {
+  return this->geom_model_.collisionPairs.size();
+}
+
+
+bool Model::is_geometry_model_initialized() {
+  return !this->geom_model_.collisionPairs.empty();
+}
+
+bool Model::check_collision(const state_representation::JointPositions& joint_positions) {
+    if (!this->is_geometry_model_initialized()) {
+        throw robot_model::exceptions::CollisionGeometryException("Geometry model not loaded for " + this->get_robot_name());
+    }
+
+    Eigen::VectorXd configuration = joint_positions.get_positions();
+
+    pinocchio::computeCollisions(this->robot_model_, this->robot_data_, this->geom_model_, this->geom_data_, configuration, true);
+
+    for(size_t pair_index = 0; pair_index < this->geom_model_.collisionPairs.size(); ++pair_index) {
+        const auto& collision_result = this->geom_data_.collisionResults[pair_index];
+        
+        if(collision_result.isCollision()) {
+            return true;
+        }
+    }
+    return false;
 }
```

Comment:
A lot of the code that you added has indentation with 4 spaces, could you make sure its 2 spaces everywhere?